File: 1143-longest-common-subsequence/1143-longest-common-subsequence.c
int longestCommonSubsequence(char* text1, char* text2) {
    int i,j;
    int len_x = strlen(text1);
    int len_y = strlen(text2);
    int** LCS = (int**)malloc((len_x + 1) * sizeof(int*));
    for (int i = 0; i <= len_x; i++) {
        LCS[i] = (int*)malloc((len_y + 1) * sizeof(int));
    }
    for (i = 0; i <= len_x; i++) {
        LCS[i][0] = 0;
    }
    for (j = 0; j <= len_y; j++) {
        LCS[0][j] = 0;
    }
    for (i = 1; i <= len_x; i++) {
        for (j = 1; j <= len_y; j++) {
            if (text1[i - 1] == text2[j - 1]) {
                LCS[i][j] = LCS[i - 1][j - 1] + 1;
            }
            else{
                if (LCS[i - 1][j] > LCS[i][j - 1]) {
                    LCS[i][j] = LCS[i - 1][j];
                } 
                else {
                    LCS[i][j] = LCS[i][j - 1];
                }
            }
        }
    }
    int result = LCS[len_x][len_y];
    for (int i = 0; i <= len_x; i++) {
        free(LCS[i]);
    }
    free(LCS);

    return result;
}

File: 1143-longest-common-subsequence/1143-longest-common-subsequence.py
class Solution:
    def longestCommonSubsequence(self, text1: str, text2: str) -> int:
        len_x, len_y = len(text1), len(text2)
    
        text1, text2 = text2, text1
        len_x, len_y = len_y, len_x
    
        LCS = [0] * (len_x + 1)
        for j in range(1, len_y + 1):
            prev = 0
            for i in range(1, len_x + 1):
                temp = LCS[i]
                if text1[i - 1] == text2[j - 1]:
                    LCS[i] = prev + 1
                else:
                    LCS[i] = max(LCS[i], LCS[i - 1])
                prev = temp
        return LCS[len_x]
        